### DISCUSSION TITLE
Scheduled Export: launchd invokes shell script with macOS permission granted (in place of crontab)

### DISCUSSION THREAD

**Discussion by rajscode** [RhetTbull/osxphotos#1080] [Show and tell]
Hopefully this might be helpful to others who want schedule a job for an osxphotos export. 

Been using osxphotos export --download-missing --use-photokit for a while with a shell script.  Has worked great when kicking off manually.  Not so great when trying with user crontab under macOS Ventura. Would get: 

` could not get authorizaton to use Photos: auth_status = 2 `

Decided to instead try using launchd and launchctl as I figured I'd be able to load and unload the job on demand as the user and I'd see any additional permission popups potentially. 

There are three locations with launchd where you can put a job plist config file:

1. /Library/LaunchDaemons - plist config file here if your job needs to run even when no users are logged in.
2. /Library/LaunchAgents - plist config file here if your job is only useful when users are logged in. Will be run as 'root' after a system reboot.
3. $HOME/Library/LaunchAgents - plist config file in this folder if the job is only useful when that user is logged in. Your job will be run under your username.

I used the 3rd option.  First I created a shell script and saved it to:

**$HOME/bin/photos_library_export_to_disk.sh**
```shell
#!/bin/sh
date
pwd
printenv
cd /Volumes/WDSSD2TB/photo_library_export
# in my situation I am using an external 2TB SSD called WDSSD2TB
$HOME/bin/osxphotos export /Volumes/WDSSD2TB/photo_library_export/photo_library_backup/ --download-missing --use-photokit --load-config /Volumes/WDSSD2TB/photo_library_export/osxphotos_backup.toml
``` 

I tested out the shell script by running it within my Terminal and granting permission for any popups that the OS generated. Once I knew the script was working properly and exporting photos I moved on to creating a launchd plist job file. 

Check to see you have a $HOME/Library/LaunchAgents directory.  If it doesn't exist for your user account, create the LaunchAgents directory. 

Create a text file that contains the XML contents below and save it to:
**$HOME/Library/LaunchAgents/com.userphotobackup.PhotosLibraryExportToDisk.plist**
```html
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
	<key>Label</key>
	<string>com.userphotobackup.PhotosLibraryExportToDisk</string>
	<key>Program</key>
	<string>/Users/userphotobackup/bin/photos_library_export_to_disk.sh</string>
	<key>RunAtLoad</key>
	<true/>
	<key>StandardErrorPath</key>
	<string>/tmp/com.userphotobackup.PhotosLibraryExportToDisk.stderr</string>
	<key>StandardOutPath</key>
	<string>/tmp/com.userphotobackup.PhotosLibraryExportToDisk.stdout</string>
</dict>
</plist>

``` 

The plist above, once loaded, will invoke the shell script when it is first loaded and run. 

From your terminal issue the following to have the launchd daemon load above plist config file:
```shell
launchctl load $HOME/Library/LaunchAgents/com.userphotobackup.PhotosLibraryExportToDisk.plist
``` 

A couple things to keep an eye on:

- Any alerts / banners / pop ups by macOS asking to authorize permission. Grant it. Also make sure that the alert popup isn't buried behind other windows by using the F3 key to quick scan all display items. 
- Monitor the two files below which is where the shell scripts output is being directed into:
  /tmp/com.userphotobackup.PhotosLibraryExportToDisk.stderr
  /tmp/com.userphotobackup.PhotosLibraryExportToDisk.stdout
  You can tail -F the files in another terminal window to keep an eye. 

Once you know everything is working as expected we can STOP the job from running (or if it's gone haywire and you need to make changes, grant priveleges, etc..).  From the terminal issue the following command:

```shell
launchctl unload $HOME/Library/LaunchAgents/com.userphotobackup.PhotosLibraryExportToDisk.plist
``` 

NOTE: When you run the job under launchd control via the launchctl load command above you will see multiple popups occur for permission authorization.  I ended up having the grant the following priveleges at various stages from start to end of everything. 

Privacy & Security:
- Files and Folders:
  - python3
    - Removabale Volumes
- Full Disk Access:
  - iTerm
  - osxphotos
  - launchd
  - sh
- Photos
  - iTerm
  - python3

UPDATED INFO: After you do everything above and below and it's all working and a month later it stops working and you see: Permission Error 1 or something similiar you've most likely updated your Python3 install (i.e. brew update / brew upgrade).  In that situation you are going to need to manually repermission for the newer python3 install (i.e python@3.9 upgraded to python@3.11) in two areas: (1) Files & Folders - Removable Volumes (2) Photos for the Privacy & Security.  You'll end up having two instance of python3 showing up there.  

UPDATED INFO No. 2 - See this comment further below in making some decisions: https://github.com/RhetTbull/osxphotos/discussions/1080#discussioncomment-8183742
<img width="468" alt="Screenshot 2023-06-17 at 2 39 08 AM" src="https://github.com/RhetTbull/osxphotos/assets/99123253/7b7b5944-846f-4611-9255-008ac450e4d7">
<img width="492" alt="Screenshot 2023-06-17 at 2 39 56 AM" src="https://github.com/RhetTbull/osxphotos/assets/99123253/b297fc05-765d-40c6-8e64-9c5ddb0ffcca">

Once the osxphotos export is working properly via launchd plist job you can now modify that plist job file so it doesn't run on load but instead ONLY runs at a particular time / day.  i.e. every Friday at 12:05AM. The updated plist file would look like:

```html
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
	<key>Label</key>
	<string>com.userphotobackup.PhotosLibraryExportToDisk</string>
	<key>Program</key>
	<string>/Users/userphotobackup/bin/photos_library_export_to_disk.sh</string>
	<key>StandardErrorPath</key>
	<string>/tmp/com.userphotobackup.PhotosLibraryExportToDisk.stderr</string>
	<key>StandardOutPath</key>
	<string>/tmp/com.userphotobackup.PhotosLibraryExportToDisk.stdout</string>
	<key>StartCalendarInterval</key>
	<array>
		<dict>
			<key>Hour</key>
			<integer>0</integer>
			<key>Minute</key>
			<integer>5</integer>
			<key>Weekday</key>
			<integer>5</integer>
		</dict>
	</array>
</dict>
</plist>
``` 

ps: There is a GUI based tool to manage launchd and plists called [LaunchControl.](https://www.soma-zone.com/LaunchControl/).  
(Not affiliated with them. Just an FYI. It costs 21 dollars.)

**Comment by RhetTbull** [comment]
Thank you for the excellent write-up!

I wonder if it might be useful to have an osxphotos command that would create the plist, the shell script, etc for the user automatically?

**Comment by oPromessa** [comment]
I've gone the crontab approach ;)

* Adjust the location of the `runOSxPhotos.cron` file.
```sh
$ crontab -l
44 3 * * sat,sun,mon /Users/MSP/Apps/runOSxPhotos.cron
```

I do have to setup under Settings -> Energy Saver -> a power up schedule  3:43 am

and then my `runOSxPhotos.cron` script runs osxphotos with the `caffeinate` command to prevent the Mac from going to sleep. Here's the runOSxPhotos.cron script:


```sh
#!/bin/bash

# Look for entries named **SET** to adjust to your configuration.
if [ -x /usr/libexec/path_helper ]; then
        eval `/usr/libexec/path_helper -s`
fi

# set verbose level
__VERBOSE=5

declare -a LOG_LEVELS
# https://en.wikipedia.org/wiki/Syslog#Severity_level
LOG_LEVELS=([0]="emerg" [1]="alert" [2]="crit" [3]="err" [4]="warning" [5]="notice" [6]="info" [7]="debug")
function .log () {
  local LEVEL=${1}
  shift
  if [ ${__VERBOSE} -ge ${LEVEL} ]; then
    echo "[${LOG_LEVELS[$LEVEL]}]" "$@"
  fi
}

# Source .bash_login
.log 7 Before source .bash_login PATH=$PATH
# **SET** To get all settings.
source ~/.bash_login
.log 7 After source .bash_login PATH=$PATH

# Activate Python Virtual Environment for osxphotos
# **SET** Set the location of your Python environment
source /Users/MSP/Documents/Apps/osxphotos/venv/bin/activate

# Set locale to en_GB.UTF-8
export LANG=en_GB.UTF-8

# Initial settings
PRG=`basename $0`
# **SET** Set location of this script
WORKFOLDER=~/Apps
# **SET** Set location of log file
FOLDER=~/Logs
LOGFILE=${FOLDER}/${PRG}.err
# **SET** Set Destination folder
DSTFOLDER=/Volumes/photo/Family.Photos
albumfile="/Users/Shared/Pictures/iPhoto Shared Library.photoslibrary/database/Photos.sqlite"

cd ${WORKFOLDER}
.log 7 CWD= `pwd`

.log 5 --- Start `date`| tee -a ${LOGFILE}

.log 5 Check if Filesystem "${DSTFOLDER}" is available | tee -a ${LOGFILE}
if [ -d "${DSTFOLDER}" ]; then
    .log 5 Filesystem available "${DSTFOLDER}"! Continuing... | tee -a ${LOGFILE}

    if [ -f "${albumfile}" ]
    then
	    .log 5 "Continuing..."
# **SET** Set location of TOML file
	    caffeinate -s osxphotos export "${DSTFOLDER}" --exportdb ${FOLDER}/.osxphotos_export.db --load-config ${WORKFOLDER}/colibri.toml >> ${LOGFILE} 2>&1
	    .log 5 Result: $? | tee -a ${LOGFILE}
    else
        .log 5 photos.db file not found. Not running.
        .log 5 photos.db file not found. Not running. | tee -a ${LOGFILE}
    fi
else
    .log 5 Folder "${DSTFOLDER}" not available! | tee -a ${LOGFILE}
fi

.log 5 --- End `date`| tee -a ${LOGFILE}
.log 5 --------- Tail ERR file... link to file: ${LOGFILE}
tail -50 $LOGFILE
.log 5 ---------
```

**Reply by rajscode** [reply]
> I've gone the crontab approach ;)

I tried with crontab using the same shell script but couldn't get it to cooperate with permissions. 
 
Noticed the following so I switched over to launchd:

```shell
man crontab
``` 

... (Darwin note: Although cron(8) and crontab(5) are officially supported under Darwin, their functionality has been absorbed into launchd(8), which provides a more flexible way of automatically executing commands.  See launchctl(1) for more information.) ...

Am going to add to the shell script to pause Blackblaze whilst the export is running. 
```shell
/Library/Backblaze.bzpkg/bztransmit -pausebackup
``` 

And to resume after the export completes:
```shell
/Library/Backblaze.bzpkg/bztransmit -completesync
``` 

My concern is that I think the -pausebackup only pauses it for 2hrs. My export typically runs approx 3-5 hrs.  So I may actually create a seperate launchd job that handles keeping backblaze paused whilst osxphotos export is active: 

12:04am every Friday - Pause for 2 hrs
Every 1hr59m onward 
if python3 osxphotos still active process
  pause backblaze again
else
  completesync

    ¯\_(ツ)_/¯

**Reply by rajscode** [reply]
@oPromessa Thanks for sharing your cron job script.

> 	    caffeinate -s osxphotos export "${DSTFOLDER}" --exportdb ${FOLDER}/.osxphotos_export.db --load-config ${WORKFOLDER}/colibri.toml >> ${LOGFILE} 2>&1
> 	    .log 5 Result: $? | tee -a ${LOGFILE}

Does your .toml config specify --download-missing --use-photokit?

Also are you on macOS 13 or an earlier version ?

Curiosity still getting me re: permission issue I was having.  I beleive (bad memory) my crontab was working under macOS 12 but I wasn't using at that point the --download-missing --use-photokit parameters. Around the time I updated to macOS 13 I added in the --download-missing --use-photokit and from that point I knew it wasn't working when run from crontab control. 

Thx!

**Reply by oPromessa** [reply]
> Does your .toml config specify --download-missing --use-photokit?
> 
Hmmm. No.
* Note 1: As I export to a NAS Synology folder I use `--exportdb` to keep the DB locally for faster access.
* Note 2: All my files are on the local Mac. Not iCloud!
* Note 3: did not had the need to use-photokit

My TOML file:
```toml
[export]
cleanup = true
db = "/Users/Shared/Pictures/iPhoto Shared Library.photoslibrary"
description_template = "Album:{album,}{newline}Description:{descr,}"
directory = "{album|filter(startswith 1|2)[/,.|:,.]}"
exiftool = true
exiftool_merge_keywords = true
exiftool_merge_persons = true
keyword_template = [
    "{keyword}",
    "{label}",
    "{searchinfo.activity?activity:{searchinfo.activity},}",
    "{searchinfo.venue_type?venue_type:{searchinfo.venue_type},}",
]
no_progress = true
not_hidden = true
ramdb = true
retry = 2
skip_bursts = true
skip_live = true
skip_original_if_edited = true
strip = true
timestamp = true
touch_file = true
update = true
verbose = true
```

> Also are you on macOS 13 or an earlier version ?
On 12.6.6
```$ sw_vers 
ProductName:	macOS
ProductVersion:	12.6.6
BuildVersion:	21G646
```

**Reply by oPromessa** [reply]
> Darwin note: Although cron(8) and crontab(5) are officially supported under Darwin, their functionality has been absorbed into launchd(8), which provides a more flexible way of automatically executing commands. See launchctl(1) for more information.) ...

- Good point. Will try out launchd to go about migrating.

**Comment by pablosed** [comment]
@rajscode this has been incredibly helpful after a number of hours of banging my head against the wall trying to achieve the exact same. Thank you for sharing!!!

@RhetTbull I strongly support you adding whatever functionality would help people achieve this in osxphotos - I am sure I wont be the last person to come along trying to do incremental unattended backups!

**Reply by pablosed** [reply]
@rajscode - back again so soon, sadly. I am wondering. Is there anything that needs to be done after changing the plist to use <StartCalendarInterval> ? I wanted mine to fire daily - so this is what I put in (basically used yours but removed the <weekday> key)

```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
        <key>Label</key>
        <string>com.userphotobackup.PhotosLibraryExportToDisk</string>
        <key>Program</key>
        <string>/Users/photobackup/.local/bin/exportpics.sh</string>
        <key>StandardErrorPath</key>
        <string>/tmp/com.userphotobackup.PhotosLibraryExportToDisk.stderr</string>
        <key>StandardOutPath</key>
        <string>/tmp/com.userphotobackup.PhotosLibraryExportToDisk.stdout</string>
        <key>StartCalendarInterval</key>
        <array>
                <dict>
                        <key>Hour</key>
                        <integer>15</integer>
                        <key>Minute</key>
                        <integer>55</integer>
                </dict>
        </array>
</dict>
</plist>
```

Unfortunately it did not fire at 15:55 today (it worked perfectly when I used the <RunAtLoad> key).

After it didnt fire the first time I tried I ran `launchctl list` and noticed that there was no relevant entry in there, so I ran `launchctl enable gui/503/com.userphotobackup.PhotosLibraryExportToDisk` - sadly still no dice. Not sure what I am missing here and hoping you can shine a light :) Thanks!

**Reply by pablosed** [reply]
Ok. So I solved my issue. Just due to total lack of understanding on my end. Was loading the plist using sudo and “bootstrap”, as soon as I did it as logged my local user and “load” it started working. Thanks again for the excellent guide.

**Comment by rajscode** [comment]
> There are three locations with launchd where you can put a job plist config file:
> 
> 1. /Library/LaunchDaemons - plist config file here if your job needs to run even when no users are logged in.
> 2. /Library/LaunchAgents - plist config file here if your job is only useful when users are logged in. Will be run as 'root' after a system reboot.
> 3. $HOME/Library/LaunchAgents - plist config file in this folder if the **job is only useful when that user is logged in. Your job will be run under your username.**
> 
> I used the 3rd option. 

Glad to read you got it going!  

Key part is you are setting it up as type 3 above and that is under the local user control that was logged in when it was set up.

I personally run two different accounts on my Mac and when I need to modify my LaunchAgent (plist) I log in as that user account (2nd account in my case) and make the changes. 

Reasons as to why I run two different accounts on the Mac are detailed at: https://forums.macrumors.com/threads/backup-solution-icloud-full-photo-library-on-external-drive-whilst-optimized-reduced-version-on-internal-drive.2369845/

**Comment by pablosed** [comment]
> I personally run two different accounts on my Mac and when I need to modify my LaunchAgent (plist) I log in as that user account (2nd account in my case) and make the changes.

@rajscode thank you for your kind response, and for the link to your (very detailed) post. It seems that the method detailed in your post would function in lieu of @RhetTbull fine piece of software? I personally have a second mac mini running, that has an optimised library on it. I then use OSXPhotos to export on a daily basis to the internal zfs mirrored drives I have setup inside a linux box i built a while back. These are shared on my local network via SMB. I then use Duplicati in a Docker container to create an encrypted backup the photos (and all the other files I store there) to Backblaze B2. Total overkill probably but its 3-2-1 backup and makes me sleep better at night :-) I am now really happy with my setup thanks to both you and @RhetTbull for getting the missing piece of my particular puzzle (exporting the originals to a flat folder structure). All of this makes me strangely happy!

Regardless, thanks again - I am always very appreciative of folk like you who take the time and effort to post their solutions online for others to use/modify!

**Comment by rajscode** [comment]
After a homebrew upgrade/update and/or OSXPHOTOS version upgrade, where an updated version is installed two types of errors may occur on a regular basis (1) OSXPHOTOS fails to run with python errors (2) Permission errors for MacOS Privacy and Security occur when executing OSXPHOTOS via a launchd job

Commands typically used to update OSXPHOTOS version and/or fix OSXPHOTOS when a python update has occurred via homebrew
```shell
pipx reinstall osxphotos  
        OR
pipx reinstall-all
```
If you want to avoid re-doing (2) Privacy and Security settings for each python 3.x release that Home Brew automatically updates you could specify which python3 version OSXPHOTOS continues to use.  Issue the following commands in place of the one above when updating OSXPHOTOS version (can specify any major installed version listed as links in: ls -la /opt/homebrew/bin/python* )
i.e. python3.11
```shell
pipx uninstall osxphotos
pipx install --python python3.11 osxphotos
```
To verify pipx installed for osxphotos properly and which version of python is being used check via the following:  
```shell
pipx list
```
To check which python3 version pipx uses by default do: 
```shell
pipx environment 
```
Your machine probably has multiple python3 versions installed if using homebrew. Each time you run a homebrew update / hombrew upgrade cycle python3 and/or pipx will get updgraded to the latest version. 

NB: If you do the above, recommend every so often for security purposes and/or OSXPHOTOS requirement purposes you update pipx to specify a newer version of Python3.  

> UPDATED INFO: After you do everything above and below and it's all working and a month later it stops working and you see: Permission Error 1 or something similiar you've most likely updated your Python3 install (i.e. brew update / brew upgrade). In that situation you are going to need to manually repermission for the newer python3 install (i.e python@3.9 upgraded to python@3.11) in two areas: (1) Files & Folders - Removable Volumes (2) Photos for the Privacy & Security. You'll end up having two instances (or more) of python3 showing up there. 

<img alt="Screenshot 2023-06-17 at 2 39 08 AM" width="468" src="https://private-user-images.githubusercontent.com/99123253/246592020-7b7b5944-846f-4611-9255-008ac450e4d7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDU2ODAwNjcsIm5iZiI6MTcwNTY3OTc2NywicGF0aCI6Ii85OTEyMzI1My8yNDY1OTIwMjAtN2I3YjU5NDQtODQ2Zi00NjExLTkyNTUtMDA4YWM0NTBlNGQ3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTE5VDE1NTYwN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTJjYzA5MzlkYTQxMTYzZWZiNDc2YzJiZGEyYmE3YzJkOTg1YmNlNDFmNjY0NDFjODI5NmJkNjNjMTE0N2RkNDQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.4KDeRzXf2QIAVwJAu7nlennkPfa6r6rAoUEyIyqX_Gw"> <img alt="Screenshot 2023-06-17 at 2 39 56 AM" width="492" src="https://private-user-images.githubusercontent.com/99123253/246592036-b297fc05-765d-40c6-8e64-9c5ddb0ffcca.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDU2ODAwNjcsIm5iZiI6MTcwNTY3OTc2NywicGF0aCI6Ii85OTEyMzI1My8yNDY1OTIwMzYtYjI5N2ZjMDUtNzY1ZC00MGM2LThlNjQtOWM1ZGRiMGZmY2NhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTE5VDE1NTYwN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWI5ZDU2ODZmNmI4YzAwN2ZjMTQ4Y2RiMWVkMGI2MzQ5NjliNjFiOWY2ZGUxZWJhZDM3ZmQyODNkODdkMjYzOTkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.fH7M-dnQ3jcvg4-GU3uwxAkmFGRmwtPeADJL4o5jUbw">
>

**Reply by RhetTbull** [reply]
Thanks for sharing -- I'm sure this will be helpful to others.  I am currently working on a companion app to osxphotos that will take care of the scheduling (and allow you to manage multiple scheduled jobs) via a GUI app.  This will be  a signed and notarized native Mac app that contains the osxphotos engine and  it will take care of the permissions and scheduling so you don't need to mess with launchd. I plan to make this a commercial app to help support the development of osxphotos. Haven't settled on a price but it will be very reasonable. I'll post an update to the Discussions page when it's ready for preview and invite osxphotos users to try out the preview version.

**Comment by rajscode** [comment]
UPDATE to the posts above and the method I am using now as it is the easiest way I've found in order not to have to do any future interventations after 1st time setup. 

In place of granting each updated version of Python3 access via Privacy & Security in MacOS, **grant those priveleges to an actual launcher application that you create via Automator**.  That app initiates the osxphotos export shell script. By priveledging a simple App you won't need to re-privelege every time python is updated as the Launcher app is the one who has the overall permissions now. 

> [!TIP]
> This launcher app also works for those running their osxphotos export command shell script manually from terminal as the same Privacy & Security TCC priveleges need to be granted repeatedly. So instead run the launcher app.  You can make all the changes you want to your shell script as often as you want as long as you call it using the launcher app you won't have to grant updated P&S rights. 
   
General Steps: 
(A) Create an App that is a shell script launcher via the Automator Application included in MacOS.  Your App will invoke your shell script that calls your osxphotos export command, etc.: 
 - Run Automator App
   - File / New / Application
   - Library / Utilties / Run Shell Script
   - Configure it for your shell (I use /bin/zsh)
     - Pass input: as arguements
     - For the script: 
       /bin/zsh -c "$1"
   - File Menu / Save as script_runner.app

(B) Test script_runner App and permission it with Privacy & Security one time instead of having to permission each version of Python3 when it gets updated by Homebrew for example:
 - From the terminal command line issue a command similiar to the one below where you've substituted: 
   - full path to your script_runner.app
   - full path to your shell script that invokes the osxphotos export command. 
     Command:
```shell  
     open -a $HOME/bin/script_runner.app $HOME/bin/photos_library_export_to_disk.sh
```
  - As the command above runs and progresses through the various stages of osxphotos being loaded and run, you'll get popup windows by MacOS asking to grant P&S permission to the script_runner.app for the following: 
    - Files and Folders / Removabale Volumes
    - Photos (takes a little while to come up as it'll ask once osxphotos gets to point where it is try to access a photo itself)

(C) If everything above in the manually initiated run has worked fine and your export finished properly you can now decide if you want to create a scheduled re-occuring LaunchAgent.  If yes, steps outlined in 1st post above BUT utilize the example XML below instead to craete a LaunchD job. NOTE: Please modify a few string values below appropriately: 
  - 8th line: com.myphotobackup.PhotoLibraryExportToDisk.ScriptRunner
  - 13th line: /Users/userphotobackup/bin/script_runner.app
  - 14th line: /Users/userphotobackup/bin/photos_library_export_to_disk_launchdjob.sh
    Make sure you write out the full path to the script_runner.app in 13th line and to your shell script in 14th line.  Don't use variables such as $HOME IMO. 
  - 20th line: i.e. 2 = 02 AM
  - 22nd line: i.e. 5 = 05 MIN
  - 24th line: i.e. 6 = Saturday
    This job runs every Saturday at 2:05 AM when user is logged in to call the script_runner.app which will then invoke the photos_library_export_to_disk_launchdjob.sh shell script. 
 
Create a plist text file that contains the XML contents below (refer to details in the 1st post on how and where to create this file)

```HTML
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
    <key>Disabled</key>
    <false/>
    <key>Label</key>
    <string>com.myphotobackup.PhotoLibraryExportToDisk.ScriptRunner</string>
    <key>ProgramArguments</key>
    <array>
        <string>open</string>
        <string>-a</string>
        <string>/Users/userphotobackup/bin/script_runner.app</string>
        <string>/Users/userphotobackup/bin/photos_library_export_to_disk_launchdjob.sh</string>
    </array>
    <key>StartCalendarInterval</key>
    <array>
        <dict>
            <key>Hour</key>
            <integer>2</integer>
            <key>Minute</key>
            <integer>5</integer>
            <key>Weekday</key>
            <integer>6</integer>
        </dict>
    </array>
</dict>
</plist>
```
> [!NOTE]
> To clean up no longer relevant Privacy & Security Settings Items such as older python3 versions referenced but not removeable via the UI, don't use tccutil but instead look into carefully deleting entries from the TCC DB directly via SQL since most of the tcc entries for this purpose are in the user space. 
> **EVERYTHING BELOW IS OPTIONAL and does NOT need to be done.** 

Step A: Go to System Settings / Privacy & Security / Photos
 - Look for mulitiple instances of python3
 - Right click on each one and select Show in Finder
 - If Finder Opens up and shows location that is good. Close Finder
 - If Finder does NOT open up it means this instance being show is no longer useful and can be removed.  Turn the item OFF
Step B: Go to System Settings / Privacy & Security / Files and Folders
 - repeat same sub steps as step A

> [!CAUTION]
> This worked for me on MacOS Sonomoa 14.3.1.  
> I don't know if it'll work for you on your machine and setup and MacOS version. 
> Unless you are VERY comfortable with risk of bricking your login instance since we are deleting items from the TCC Security database I recommend
> you do NOT do this 

Step C:  Removing the old and no longer useful or needed python3 entries that we turned off above in steps 1 and 2: 
 - Locate all entries in user space TCC DB where item is disabled and bundleid (client) is python. (NOTE: Keep an eye on the values for service as in this case in particular we are looking only for kTCCServicePhotos and kTCCServiceSystemPolicyRemovableVolumes.  If you see other values you may want to consider modifying SQL delete command below to include an additional condition in the where statement for service to match the value you are looking to delete explcitly, so something additional isn't deleted.
```SQL
   sqlite3 ~/Library/Application\ Support/com.apple.TCC/TCC.db "select client, auth_value, service from access where auth_value='0' and client like '%python%' order by client"
```
 - Take each unique client from list above and delete it from the user space TCC DB for entries in access table where auth_value=0 and client=what we entered from above. i.e.:
```SQL
   sqlite3 ~/Library/Application\ Support/com.apple.TCC/TCC.db "delete from access where auth_value='0' and client='/opt/homebrew/Cellar/python@3.11/3.11.5/Frameworks/Python.framework/Versions/3.11/bin/python3.11'"
```
   NOTE: Don't execute command above as is.  You must modify the client='' portion.  Make sure you maintain the proper quotes symbols. Also please remember prior comment re: service values
   Repeat as needed for each unique client from prior step.

**Reply by RhetTbull** [reply]
Thank you @rajscode ! Very useful and thorough rundown.

**Reply by rajscode** [reply]
I suspect the work you doing on [applecrate](https://github.com/RhetTbull/applecrate) will obviate the need for above in regards to osxphotos and anyone else who adopts applecrate for their command line tools.  Was more so just documenting things so others can find via a search in case its needed as the TCC stuff is pretty poorly documented by Apple IMO and very limited in access via the tccutil.

**Reply by jasonhollis** [reply]
This work is simply awesome!  @RhetTbull RheTbull pointed me here this morning in what was hopefully not a completely annoying email spam session about various things.  It's funny I was thinking about AppleScript or even Shortcuts to launch it.  @rajscode have you played with Shortcuts?  I will have a look today.  Assuming you could get the environment/path set properly, I think it would be very easy to handle the schedule, etc.  I'm using a free add-in for Shortcuts called Actions ... I have not investigated it all yet but I suspect Actions will facilitate this in Shortcuts much more easily [Actions](https://apps.apple.com/au/app/actions/id1586435171) This is the GitHub for [further actions](https://github.com/sindresorhus/Actions#looking-for-more)I'm going to have a deeper look at applecrate as well.

**Comment by odedia** [comment]
Thank you @rajscode for this idea!
I'm facing the same `auth_status=2` error, when I launch a backup via Arq Backup. Arq lets you execute a script before or after the backup. Therefore, I need the script to block and not run in the background in order for Arq to know when it's completed successfully.
I tried running Script Runner and it seems to be running in the background and with logs output to /tmp with `permission denied` message.
Can you think of an idea to block the process and redirect output somewhere else?
Thanks again.

**Comment by odedia** [comment]
Finally figured it out. Based on @rajscode guide above, I created an App called ScriptRunner.app.
I then created a script called arqagent.sh that has the following:

```
#!/bin/bash
echo ------------------------------------------------------
echo Running osxphotos and rsync, this could take a while. 
echo You can monitor /tmp/output.txt if you want.
echo ------------------------------------------------------

open --wait-apps -a /Applications/ScriptRunner.app /Users/odedia/rsync.sh
cat /tmp/output.txt
rm /tmp/output.txt
```

`--wait-apps` will basically block the command line until the app exits, which will happen at the end of the script run. It's important because I want the backup to start _after_ osxphotos is completed.

The second part is the log file - for the life of me I couldn't get the console output to show up when running the scipt through an automator app, so I basically had all lines printed with `>> /tmp/output.txt` in `rsync.sh`. I want Arq to show the output of the execution so I can monitor if the backup worked as expected.
Even here - I could not get the basic logs to show up unless I added `--verbose` to the osxphotos parameters.

The third thing I had to do was give ScriptRunner.app Full Disk Access. Otherwise, I would get a popup similar to this one https://github.com/microsoft/vscode/issues/201087 on every run. Ofcourse I also need to give ScriptRunner access to Apple Photos but that is preserved between runs and I don't get that annoying popup on every single run.

All in all - it works great! :D


**EDIT:** Here's a better implementation that prints to console output in real time and not just at the end of the execution (thanks ChatGPT):
```
#!/bin/bash
echo ------------------------------------------------------
echo Running osxphotos and rsync, this could take a while. 
echo Tailing log at /tmp/output.txt
echo ------------------------------------------------------

rm -f /tmp/output.txt
touch /tmp/output.txt
chown odedia:staff /tmp/output.txt

# Function to cleanup child processes
cleanup() {
    echo "Stopping child processes..."
    if [ -n "$tail_pid" ]; then
        echo "Terminating cat process with PID $tail_pid..."
        kill $tail_pid
    fi
    pkill -P $$
}


# Trap exit signals to call cleanup function
trap cleanup EXIT

# Execute cat command in the background and save its PID
tail -f /tmp/output.txt &
tail_pid=$!

open --wait-apps -a /Applications/ScriptRunner.app /Users/odedia/rsync.sh
scriptRunnerStatus=$?

# Check if ScriptRunner.app executed successfully
if [ $scriptRunnerStatus -ne 0 ]; then
    echo "ScriptRunner failed with exit status $scriptRunnerStatus"
    exit $scriptRunnerStatus
fi

kill $tail_pid
```

**Reply by jasonhollis** [reply]
Folks I reckon the simplest way to get everything you want is via shortcuts.  I have mine kicking back pretty verbose output and building me a little results log on the local Mac.  I'm then using the plist method above to launch the shortcut and all is groovy except now I'm getting an error that I can't seem to get around with Osxphotos.  That I will post separately.
